README.md
# arp-sigils

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar arp-sigils-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2022 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2022-12-30
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2022-12-30
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/arp-sigils/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/arp-sigils/compare/0.1.0...0.1.1


test/arp_sigils/core_test.clj
(ns arp-sigils.core-test
  (:require [clojure.test :refer :all]
            [arp-sigils.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/arp_sigils/glyphs.clj
(ns arp-sigils.glyphs
  (:require [clojure2d.core :as c2d]
            [fastmath.core :as fm])
  )

(defn scale [x]
  (* 10 x))



(defn attach-glyph [base addition]
  (let [acount (count (:attach base))]
    (assoc base :subglyphs (vec (repeat acount addition)))))

(defn attach-out-glyph [base addition]
  (assoc base :outglyph addition))

(defn attach-glyphs-by-line [a b]
  (let [join (join-line)]
    (attach-out-glyph a (attach-out-glyph join b))))


(defn size-glyph [glyph]
  (let [theta     (:rotation glyph)
        [x y]     (:in glyph)
        bb        (:bbox glyph) 
        subglyphs (:subglyphs glyph)
        sizefn    (:sizefn glyph)]
    (if subglyphs 
      (let [sizedsubs (map size-glyph subglyphs)
            sized     (sizefn glyph sizedsubs)]
        (merge glyph sized {:subglyphs sizedsubs}))
      glyph
      )))

(defn rotate-shape-at [shape x y theta]
    (.createTransformedShape (java.awt.geom.AffineTransform/getRotateInstance  theta x y) 
    shape))


;(attach-rect2d-at (c2d/rect-shape -10 -10 20 20) 0 -10 -10 0 fm/-HALF_PI)
(defn attach-rect2d-at [rect inx iny ax ay theta]
  (let [tform (doto (java.awt.geom.AffineTransform.)  
                    (.translate (- inx) (- iny))
                    (.rotate theta)
                    (.translate ax ay)
                )
        path  (.createTransformedShape tform rect)
        ]
    (.getBounds path)
    )
    )

(defn bbox->wh [bbox]
  (let [[x1 y1 x2 y2] bbox] 
    [(- x2 x1) (- y2 y1)]))

(defn rect2d->corners [rect]
  [(.getX rect) 
   (.getY rect)
   (+ (.getX rect) (.getWidth rect))
   (+ (.getY rect) (.getHeight rect))
   ])

(defn union-bboxes [& bboxes]
  (->> bboxes
    (mapv (partial apply c2d/rect-shape) , )
    (reduce #(.createUnion %1 %2) ,)
    rect2d->corners))

(defn join-line []
  {:name :join
   :parts [[:line -10 0 10 0]]
   :width 20
   :in [-10 0]
   :out [10 0]})

(defn zero []
  {:name :zero
   :parts [[:arc 0 0 20 20 0.0 (* 2.0 Math/PI)]]
   :subglyphs nil
   :outglyph nil
   :out [10 0]
   :width 20
   :bbox (c2d/rect-shape -10 -10 10 10)
   :in [-10 0]
   :attach [[0 0 0.0]]
   :sizefn (fn [self bboxes] 
             (let [bbox    (-> bboxes first :bbox)
                   [xs ys] (rect2d->corners bbox)]
               (merge self {:bbox bbox
                            :parts [[:arc 0 0 xs ys 0.0 fm/TWO_PI]]})))})

(defn one []
  {:name :one
   :parts  [[:line 0 -20 0 20] [:point 0 -20]]
   :width 20
   :bbox (c2d/rect-shape -10 -10 10 10)
   :in [-10 0]
   :attach [[0 20 fm/HALF_PI]]
   :sizefn (fn [self bboxes]
             (let [bbox    (-> bboxes first :bbox )
                   [xs ys] (bbox->wh bbox)]
               (merge self {:bbox bbox
                            :parts [[:arc 0 0 xs ys 0.0 fm/TWO_PI]]})

               )
             )
  })

(defn two []
  {:name :two
   :parts  [[:line 0 -20 0 20] [:point 0 -20] [:point 0 20]]
   :out [10 0]
   :width 20
   :bbox (c2d/rect-shape -10 -10 10 10)
   :in [-10 0]
   :attach [[0 20 fm/HALF_PI] [0 -20 fm/-HALF_PI]]
   :sizefn (fn [bboxes])
  })

(defn three []
  {:name :three
   :parts [[:line 0 0 0 20] 
           [:line 0 0 -10 -20] [:point -10 -20]
           [:line 0 0 10 -20] [:point 10 -20]
           ]
   :width  20
   :out [10 0]
   :bbox (c2d/rect-shape -10 -10 10 10)
   :in [-10 0]
   :attach [[-10 -20 0 (+ fm/PI fm/QUARTER_PI)]
            [10 -20 (+ fm/QUARTER_PI)]]
   :sizefn (fn [boxes])
   })

;{:main [:line :one :line :zero :line]}

;"94:c6:91:16:48:49"
;{:main [:line :nine [:four] :line :c [:six] :line :nine [:one] ]}



src/arp_sigils/sigils.clj
(ns arp-sigils.sigils
  (:require [clojure2d.core :as c2d]
            [fastmath.core :as fm])
  )

(defn scale [x]
  (* 10 x))



(defn attach-sigil [base addition]
  (let [acount (count (:attach base))]
    (assoc base :subsigils (vec (repeat acount addition)))))

(defn size-sigil [sigil]
  (let [theta     (:rotation sigil)
        [x y]     (:in sigil)
        bb        (:bbox sigil) 
        subsigils (:subsigils sigil)
        sizefn    (:sizefn sigil)]
    (if subsigils 
      (let [sizedsubs (map size-sigil subsigils)
            sized     (sizefn sigil sizedsubs)]
        (merge sigil sized {:subsigils sizedsubs}))
      sigil
      )))

(defn rotate-shape-at [shape x y theta]
    (.createTransformedShape (java.awt.geom.AffineTransform/getRotateInstance  theta x y) 
    shape))


(attach-rect2d-at (c2d/rect-shape -10 -10 20 20) 0 -10 -10 0 fm/-HALF_PI)
(defn attach-rect2d-at [rect inx iny ax ay theta]
  (let [tform (doto (java.awt.geom.AffineTransform.)  
                    (.translate (- inx) (- iny))
                    (.rotate theta)
                    (.translate ax ay)
                )
        path  (.createTransformedShape tform rect)
        ]
    (.getBounds path)
    )
    )

(defn bbox->wh [bbox]
  (let [[x1 y1 x2 y2] bbox] 
    [(- x2 x1) (- y2 y1)]))

(defn rect2d->corners [rect]
  [(.getX rect) 
   (.getY rect)
   (+ (.getX rect) (.getWidth rect))
   (+ (.getY rect) (.getHeight rect))
   ])

(defn union-bboxes [& bboxes]
  (->> bboxes
    (mapv (partial apply c2d/rect-shape) , )
    (reduce #(.createUnion %1 %2) ,)
    rect2d->corners))

(defn zero []
  {:parts [[:arc 0 0 20 20 0.0 (* 2.0 Math/PI)]]
   :width 20
   :bbox (c2d/rect-shape -10 -10 10 10)
   :in [-10 10]
   :rotation 0.0
   :attach [[0 0 0.0]]
   :sizefn (fn [self bboxes] 
             (let [bbox    (-> bboxes first :bbox)
                   [xs ys] (rect2d->corners bbox)]
               (merge self {:bbox bbox
                            :parts [[:arc 0 0 xs ys 0.0 fm/TWO_PI]]})

             ))
   })

(defn one []
  {:parts  [[:line 0 -20 0 20] [:point 0 -20]]
   :width 0
   :bbox (c2d/rect-shape -10 -10 10 10)
   :in [-10 10]
   :rotation 0.0
   :attach [[0 20 fm/-HALF_PI]]
   :sizefn (fn [self bboxes]
             (let [bbox    (-> bboxes first :bbox )
                   [xs ys] (bbox->wh bbox)]
               (merge self {:bbox bbox
                            :parts [[:arc 0 0 xs ys 0.0 fm/TWO_PI]]})

               )
             )
  })

(defn two []
  {:parts  [[:line 0 -20 0 20] [:point 0 -20] [:point 0 20]]
   :width 0
   :bbox [-10 -10 10 10]
   :in [-10 10]
   :rotation 0.0
   :attach [[0 20 fm/PI] [0 -20 fm/-HALF_PI]]
   :sizefn (fn [bboxes])
  })


;{:main [:line :one :line :zero :line]}

;"94:c6:91:16:48:49"
;{:main [:line :nine [:four] :line :c [:six] :line :nine [:one] ]}



src/arp_sigils/draw.clj
(ns arp-sigils.draw
  (:require [arp-sigils.glyphs :as g]
            [clojure2d.core :as c2d]
            [clojure2d.color :as color]
            [clojure2d.pixels :as pix]
            [fastmath.core :as fm]
            ))

(def window)

(def state (atom {:mode :sigil
                  :addr "94:c6:91:16:48:49"
                  :exception nil
                  :exception-timeout 0}))
(def errstate (atom {}))

(def stroke 3)

(defn reset-state []
  (reset! state {:mode :sigil
                :addr "94:c6:91:16:48:49"
                :exception nil
                :exception-timeout 0}))

(def canvas (c2d/canvas 800 400))

(defn draw-strokes [canref strokes]
                   (doseq [part (:parts strokes)
                         :let [kind   (first part)
                               params (vec (rest part))
                               ;_ (println kind params)
                               ]] 
                     (case kind
                       :line  (apply (partial c2d/line canref) params)
                       :arc   (apply (partial c2d/arc canref) params)
                       :point (do 
                                (c2d/set-stroke canref (* 2 stroke))
                                (c2d/point canref params)
                                (c2d/set-stroke canref  stroke)
                                )
                       )
                     )
                   (:attach strokes)
  )


(defn draw-glyph [canref glyph]
  (let [{:keys [parts width in out outglyph subglyphs attach]} glyph
        [inx iny] in
        ]
    (c2d/translate canref (* 0.5 width) 0)
    (c2d/set-color canref :red)
    (c2d/set-stroke canref 0.5)
    ;(c2d/arc canref 0 0 20 20 0 fm/TWO_PI)
    (c2d/set-color canref :green)
    (c2d/set-stroke canref stroke)
    (doseq [part parts
            :let [kind (first part)
                  params (rest part)]]
      (case kind
        :line  (apply (partial c2d/line canref) params)
        :point (apply (partial c2d/point canref) params)
        :arc   (apply (partial c2d/arc canref) params)
        ))
    (if subglyphs
      (do 

       (c2d/set-color canref :red)
       (c2d/set-stroke canref 5)
       ;(c2d/point canref 0 0)
      (doseq [[attachment glyph] (map vector attach subglyphs)
              :let [[x y theta] attachment]]
        (c2d/push-matrix canref)
        (c2d/translate canref x y)
        (c2d/rotate canref theta)
        (draw-glyph canref glyph)
        (c2d/pop-matrix canref)
      ))
      )
    (if outglyph
      (do 
        (apply (partial c2d/translate canref) out)
        ;(c2d/translate canref width 0)
        (draw-glyph canref outglyph)
        )
      )))

;(map :parts [(g/zero) (g/one) (g/one) (g/zero) (g/two)])
(defn draw-sigil [canvas sigil]
  (c2d/with-canvas [canref canvas]
                   (let [hw (* 0.5 (:w canref))
                         sw (- hw (* 0.5 hw))
                         sh (* 0.5 (:h canref))
                         strokespacer 20
                         ]

                     (c2d/push-matrix canref)
                     (c2d/translate canref sw sh)
                     (c2d/set-color canref :green)
                     (c2d/set-stroke canref stroke)
                     (draw-glyph canref sigil)
                     (comment loop [glyph sigils
                            ;x 0
                            ;y 0
                            ]
                       (if (nil? sigils)
                         nil
                         (let [sigil (first sigils)
                               in    (:in sigil)
                               out   (:out sigil)
                               width (:width sigil)
                               half  (* 0.5 width)]
                           (c2d/set-color canref :lime)
                           (c2d/line canref [0 0] [strokespacer 0])
                           (c2d/set-color canref :green)
                           (c2d/translate canref [(+ strokespacer half) 0])
                           (draw-strokes canref sigil)
                           (c2d/translate canref [half 0])
                           (recur (next sigils) )

                           )
                         )
                       
                       )
                     )))

(def testsigil 
  (g/attach-out-glyph 
    (g/zero)
    (g/attach-out-glyph (g/join-line) 
                        (g/attach-glyph (g/one) (g/two)))))

(def testsigil (g/size-glyph testsigil))

(defn digit->glyphfn [d]
  (case d
        \0 (g/zero)
        \1 (g/one)
        \2 (g/two)))

(def testsigil
  (arp->sigil "01:11:21:20"))

(defn arp->sigil [macstr]
  (let [rev    (clojure.string/reverse macstr)
        segs   (clojure.string/split rev #":")
        gpairs (for [seg segs
                     :let [a (first seg)
                           b (second seg)]]
                 (g/attach-glyph (digit->glyphfn b) (digit->glyphfn a)))
        lined  (map #(g/attach-out-glyph (g/join-line) %) gpairs) ]
    (reduce #(g/attach-out-glyph %) lined)))

(defn draw-testbed [canvas]
  (c2d/with-canvas [canvas canvas]
                   ;(draw-sigil canvas (g/two))
                   (c2d/set-color canvas :red)
                   (c2d/set-stroke canvas 1)
                   (c2d/translate canvas 100 100)
                   (c2d/line canvas -100 0 100 0)
                   (c2d/line canvas 0 -100 0 100)
                   (c2d/arc canvas 0 0 100 100 0 fm/TWO_PI)
                   (c2d/arc canvas 0 0 200 200 0 fm/TWO_PI)
                   (c2d/shape canvas (c2d/rect-shape -10 -10 20 20))
                   ;(c2d/shape canvas (g/attach-rect2d-at (c2d/rect-shape -10 -10 20 20) 0 -10 -10 0 fm/-HALF_PI) true)
                   
                   )
  (c2d/with-canvas [canvas canvas]
                   (draw-sigil canvas testsigil)
  ))

(defn draw-error [canvas]
  (let [exc     (:exception @state)
        message (.getMessage exc)
        traces  (.getStackTrace exc)
        lines   (filter #(re-matches #".*pftop_viz.*" %) (concat (map str traces)))
        tmout   (:exception-timeout @state)
        diff    (max 0 (- tmout (System/currentTimeMillis)))  
        ]
    (cond 
      (= 0 tmout)
      (do
        (swap! state assoc :exception-timeout (+ (* 1000 10) (System/currentTimeMillis)))
        )
      (>= (System/currentTimeMillis) tmout)
      (do
        ;(c2d/with-canvas-> canvas (c2d/line 0 0 1000 1000))
        (reset! errstate @state)
        (reset-state)
        ) 
      )
    (c2d/with-canvas [canvas canvas]
                     (c2d/set-font canvas "Hack")
                     (c2d/set-font-attributes canvas 10 :bold)
                     ;(c2d/set-background canvas 45 51 45)
                     (c2d/set-color canvas :lime)
                     (c2d/text canvas message 10 20)
                     (c2d/text canvas (str diff) (- (:w canvas) 200) (- (:h canvas) 20))
                     (doall (map-indexed #(c2d/text canvas %2 10 (+ 50 (* 10 %1))) lines))
                     (c2d/text canvas (:mode @state) (- (:w canvas) 200) (- (:h canvas) 30))
                     (c2d/text canvas (:addr @state) (- (:w canvas) 200) (- (:h canvas) 40))
                     )
    ))

(defn draw [canvas _ _ _]
  (try
    (c2d/with-canvas-> canvas 
                       (c2d/set-background :black)) 

    (case (:mode @state)
      :error (draw-error canvas)
      :sigil nil;(draw-sigil canvas)
      nil)
    (draw-testbed canvas)
    (catch Exception e
      (do
        (swap! state assoc :exception e)
        (swap! state assoc :mode :error)
        (draw-error canvas))))
 )

(defn start [fullscreen?]
  (if fullscreen?
    (let [w (c2d/screen-width)
          h (c2d/screen-height)]
      (def canvas (c2d/canvas w h))
      (def weather-canvas (c2d/canvas w h))
      (def window 
        (c2d/show-window {:canvas canvas 
                          :window-name "arp-sigils"
                          :draw-fn draw
                          :fps 30
                          :position [0 0]
                          :w w
                          :h h
                          }
                         ))
      )
    (def window 
      (c2d/show-window {:canvas canvas 
                        :window-name "arp-sigils"
                        :draw-fn draw
                        :fps 30
                        }
                       ))
    ))

(comment
  (def pi (.. (c2d/rect-shape 0 0 10 10) (getPathIterator nil) ))
  (def a (double-array 6))
  (.isDone pi)
  (.next pi)
  (.currentSegment pi a)
  (vec a)
(identity java.awt.geom.PathIterator/SEG_CLOSE)

  (let [pi (.. (c2d/rect-shape 0 0 10 10) (getPathIterator nil) )]
    (while (not (.isDone pi))
      (.currentSegment pi (double-array 6)))
      (.next pi)
      )
  (type (c2d/rect-shape 0 0 10 10))
  (clojure.reflect/reflect (c2d/rect-shape 0 0 10 10))
  (->  (c2d/rect-shape 0 0 10 10)
    (c2d/shape->path-def
    (.createTransformedShape (java.awt.geom.AffineTransform/getRotateInstance  1.0 0.0 0.0) 
    (c2d/rect-shape 0 0 10 10))
      )
  )
(doto (java.awt.geom.AffineTransform.) (.rotate 1.0) (.translate 10 10))
    (c2d/shape->path-def (c2d/rect-shape 0 0 10 10))
   (c2d/bounding-box )
    (.rotate 90)
  (->  (doto (c2d/rect-shape 0 0 10 10) (.rotate 90))
    (.rotate 90)
   (c2d/bounding-box )
  (-> (g/rotate-shape-at (c2d/crect-shape -10 -10 10 10) -10 0 Math/PI)
   c2d/bounding-box )
  (c2d/path-def->shape (:parts (g/one)))
    (c2d/path-def->shape [[:line 0.0 0.0 1.0 1.0]])
  )
  (g/size-sigil (g/zero))

  (.getBounds (.createUnion (c2d/rect-shape 0 0 10 10) (c2d/rect-shape 10 10 20 20)))
  (g/size-sigil (g/attach-sigil (g/zero) (g/one)))
  
  )
